File: 2021/day1/day1.ex
defmodule Day1 do
  def problem1() do
    read_file()
    |> count_increasing_depths()
  end

  def count_increasing_depths(depths) do
    depths
    |> Enum.chunk_every(2, 1, :discard)
    |> Enum.filter(fn [prev, next] -> next > prev end)
    |> length()
  end

  def problem2 do
    read_file()
    |> combine_depths()
    |> count_increasing_depths()
  end

  def combine_depths(depths) do
    depths
    |> Enum.chunk_every(3, 1, :discard)
    |> Enum.map(&Enum.sum/1)
  end

  defp read_file do
    "input.txt"
    |> File.read!()
    |> String.split("\n")
    |> Enum.reject(&(&1 == ""))
    |> Enum.map(&String.to_integer/1)
  end
end

File: 2021/day1/day1.erl
-module(day1).
-export([problem1/0, problem2/0]).

problem1() ->
    count_increasing_depths(read_file()).

problem2() ->
    count_increasing_depths(chunk_by(3, read_file())).

chunk_by(2, [_Hd | Tl] = List) ->
    lists:zip(lists:droplast(List), Tl);
chunk_by(3, [_Hd | [_TlHd | TlTl] = Tl] = List) ->
    lists:zip3(lists:droplast(lists:droplast(List)), lists:droplast(Tl), TlTl).

count_increasing_depths(Depths) ->
    length(lists:filter(fun({X, Y}) -> Y > X end, chunk_by(2, Depths))).

read_file() ->
    {ok, Contents} = file:read_file("input.txt"),
    Depths = lists:filter(fun(Line) -> Line /= "" end, string:split(binary_to_list(Contents), "\n", all)),
    lists:map(fun(Line) ->
                      {Int, _} = string:to_integer(Line),
                      Int
              end, Depths).
